keras_nlp/src/models/segment_anything/sam_backbone.py

```diff
@@ -0,0 +1,108 @@
+# Copyright 2024 The KerasNLP Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from keras_nlp.src.api_export import keras_nlp_export
+from keras_nlp.src.models.backbone import Backbone
+from keras_nlp.src.models.vit_det.vit_det_backbone import ViTDetBackbone
+
+
+@keras_nlp_export("keras_nlp.models.SAMBackbone")
+class SAMBackbone(Backbone):
```

Comment:
Take a look at https://github.com/keras-team/keras-nlp/pull/1820 for an idea of how we should architect this.

I think we should back `SAMBackbone` take in three things `image_encoder`, `prompt_encoder` and `mask_decoder`. Use `ViTDetBackbone` directly for the `image_encoder` and expose the other two classes for `SamPromptEncoder` and `SamMaskDecoder`.

I'm not exact sure what we should output here, but I suspect we should output a dict with all the various embeddings and just not apply the mask decoder yet. `mask_decoder` can be an unused property on the backbone. Then on the task, we simply apply the backbone's mask decoder to the embedding output.

Reply:
The general approach is that our backbone should contain all the pretrained weighs we care about, in the `model.weights.h5`. The task really just set's thing up for a specific training/inference setup.

Reply:
done!